lms/static/scripts/frontend_apps/components/BasicLtiLaunchApp.js

```diff
@@ -209,43 +202,11 @@ export default function BasicLtiLaunchApp({ rpcServer }) {
     }
   }, [authToken, canvas.authUrl, fetchContentUrl, fetchGroups]);
 
-  if (showIframe) {
-    const iFrame = (
-      <Fragment>
-        {showSpinner && <Spinner className="BasicLtiLaunchApp__spinner" />}
-        <iframe
-          // Render the iframe hidden if still loading something (such as the groups)
-          style={{
-            visibility: fetchCount === 0 ? 'visible' : 'hidden',
-          }}
-          width="100%"
-          height="100%"
-          className="js-via-iframe"
-          src={contentUrl}
-          title="Course content with Hypothesis annotation viewer"
-        />
-      </Fragment>
-    );
-
-    if (grading && grading.enabled) {
-      // Use the LMS Grader.
-      return (
-        <LMSGrader
-          students={grading.students}
-          courseName={grading.courseName}
-          assignmentName={grading.assignmentName}
-        >
-          {iFrame}
-        </LMSGrader>
-      );
-    } else {
-      return iFrame;
-    }
-  } else {
-    // Render either a Spinner or Dialog.
+  // Render
```

Comment:
I simplified the render() logic to the following

1. If there is an error , render the error
2. If there is a spinner, render the spinner
3. Else Render the iFrame

Reply:
Since it is possible to transition from state (3) to state (2) or (1), it is possible that the iframe gets created, the client loads and then it gets destroyed. There are some places in the LMS app that can end up with references to the iframe which would need to handle it being destroyed:

1. If the iframe goes away while the JSON-RPC server is handling a `requestGroups` request from it, then the attempt to `postMessage` a response back to the iframe (`event.source.postMessage(...)`) will fail in https://github.com/hypothesis/lms/blob/86923468995891023c16f869e35d50256e5f8a92/lms/static/scripts/postmessage_json_rpc/server/server.js#L79.
2. The `Server` object exposes a Promise for the sidebar window via `sidebarWindow`. Promises can only be resolved once, so if the promise gets resolved with an initial iframe and that iframe is later destroyed and a new one is created, the `sidebarWindow` Promise will continue to try to return the original iframe (which will become `null` once it is destroyed).

So to resolve this, we either need to make the code in various places in the LMS frontend cope with the iframe being destroyed and re-created, or re-work the render code here to keep the iframe around once it is rendered but just hide it.

Keeping the iframe rendered-but-hidden is going to be a better user experience because it will show up instantly when the error/authorization step is completed, rather than needing to be reloaded from scratch.

Reply:
> Something that is not clear to me yet - why did this work before? Since the error state was not being reset after `authorize()` returned previously.

Likely because `showIframe` (now removed) was true and it didnt matter that the errorState have a value in it. And in spite of that working, it seems confusing, best to clean it up